### PR TITLE
(WIP) introduce a way for users to emulate the `platform` object for local development

### LINKED ISSUE #4292: Platform context fallbacks
### Describe the problem

- When developing with `svelte-kit dev`, the `event.platform` object is always empty, with no great way to mock it.
- When building using an adapter like Cloudflare Workers, `event.platform` is empty for prerendering.

### Describe the proposed solution

Ability to provide a `platform` object in `svelte.config.js` that is
- Substituted for `event.platform` when `event.platform` is undefined. 
- Merged with `event.platform`, useful for specifying optional/default platform values.
- Both?

### Alternatives considered

Perhaps allowing a 'transform' function that accepts an event may be better, in case platform needs to change based on inputs?

### Importance

would make my life easier

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
This PR introduces an `emulatePlatform` method that adapters can provide that allows them to emulate the `platform` object for users during local development (and preview).

More specifically this is used in the Cloudflare adapters, so that (using the newly wrangler utility [`getBindingsProxy`](https://github.com/cloudflare/workers-sdk/pull/4523)) they read a `wrangler.toml` file from the user's directory and generate an appropriate platform object for it accordingly.

As an example of the use of this, see this [simple demo](https://github.com/dario-piotrowicz/svelte-kit-cf-app-with-bindings-demo) I made.

___

This PR has significantly changed from the original idea/proposal I had, just for recording it here's the original PR description:
<details>
<summary>Original `devPlatform` idea (outdated)</summary>
Hello :wave:

The changes here introduce a `devPlatform` option for the `kit` config so that developers can provide their own platform mocking implementation for local development.

Given the conversation in #4292, @rich-harris' [pushback](https://github.com/sveltejs/kit/issues/4292#issuecomment-1065208915)), #3535, #8397, and so on... I am aware that the changes here might just get rejected but I wanted to propose them anyways as I think that this could be a valid short/medium term solution that many developers could take advantage of right now (instead of making them indefinitely wait for a better solution that might or might not come at some point in the future).

From an outsider's perspective and the issues I mentioned above (which seem pretty stale) I don't know what the Svelte team's current plan is regarding this issue, if you have a different/better solution in the works that can be applied soon-ish that'd be awesome, but if you don't please give this PR's a second thought since this clearly seems to be an issue that various developers struggle with. Moreover please note that the changes here are pretty minimal so they should be very simple to deprecate/remove in the future if a better alternative does get implemented.

As an example of the use of this `devPlatform` config, I've implemented this [simple demo](https://github.com/dario-piotrowicz/svelte-kit-cf-app-with-bindings-demo) which shows how this could be used to locally develop an application using a Cloudflare KV binding, you can see [here](https://github.com/dario-piotrowicz/svelte-kit-cf-app-with-bindings-demo/blob/main/svelte.config.js#L16-L27) how the `devPlatform` is being set and [here](https://github.com/dario-piotrowicz/svelte-kit-cf-app-with-bindings-demo/blob/main/src/routes/%2Bpage.server.ts) how it is being used.

Currently the only alternative workaround that can allow developers to use bindings with the Vite dev server is by leveraging server hooks, I included an example of that [here](https://github.com/dario-piotrowicz/svelte-kit-cf-app-with-bindings-demo/blob/no-devPlatform/src/hooks.server.ts). As you can see this is much much more cumbersome than the proposed API and also effects the production build, so it's definitely not a great alternative in my opinion.

Additionally I'd like to point out that there's a precedence of such an API, QwikCity is actually doing the exact same thing in their Vite plugin (see [their docs](https://qwik.builder.io/docs/advanced/vite/#platform)), showing that even if not perfect this seems to be an acceptable solution for this problem.

</details>

___

resolves #4292
(also #4908, and maybe more)
___

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by saturnonearth** [comment]
This would be great

**Review by benmccann — CHANGES_REQUESTED** [review]
This feels like the wrong API to me. I think it'd be better if the adapter could be used to provide the `platform` in dev mode as well

**Comment by Conduitry** [comment]
Yeah, I agree, this doesn't feel like the right API. Until we figure out how we want to have adapters affect dev mode, I think the best option is to have a `handle` hook that does some assignments in an `if (dev) { ... }`, as mentioned in the PR description. This gets treeshaken out of production builds. It seems like an entirely acceptable workaround to me.

**Comment by dario-piotrowicz** [comment]
> This feels like the wrong API to me. I think it'd be better if the adapter could be used to provide the platform in dev mode as well

@benmccann are saying that you'd be ok with this approach if I moved the `devPlatform` to be part of an adaptor config instead? or are you saying that this feels like the wrong approach/API full stop?

**Comment by dario-piotrowicz** [comment]
> Yeah, I agree, this doesn't feel like the right API. Until we figure out how we want to have adapters affect dev mode, I think the best option is to have a handle hook that does some assignments in an if (dev) { ... }, as mentioned in the PR description. This gets treeshaken out of production builds. It seems like an entirely acceptable workaround to me.

As I also mentioned in the PR description, I do agree with you saying that there might be a better solution on how adapters should effect dev mode as a whole, what I am questioning here is: when is that solution going to be implemented? any time soon? if not, wouldn't it be nice to provide a valid stopgap easily-removable/deprecatable solution that can help solving the issue right now?

Regarding the `handle` solution again I kind of do agree with you that it is a kind of acceptable but it does still required developers to implement this hands-on hack which does have (even if minimal) implications in the production build (since the dev code can be tree shaken away but the handle itself cannot).

Basically what I'm trying to ask here is, are the benefits of implementing such a change outweighed by the cons?

File: pnpm-lock.yaml [hunk 1]
```diff
@@ -6609,6 +7208,33 @@ packages:
...
+    resolution: {integrity: sha512-jEnqpY+9/J4VPjtuEnS2lhCPXkvbDClnMalSWaRxSx+1tiTWMJhMjtK9oyXLdO+ZUf9Q4LvFTYSPm8O1uwmnxQ==}
+    engines: {node: '>=16.17.0'}
+    hasBin: true
+    dependencies:
+      '@cloudflare/kv-asset-handler': 0.2.0
+      '@esbuild-plugins/node-globals-polyfill': 0.2.3(esbuild@0.17.19)
+      '@esbuild-plugins/node-modules-polyfill': 0.2.2(esbuild@0.17.19)
+      blake3-wasm: 2.1.5
+      chokidar: 3.5.3
+      esbuild: 0.17.19
```
Comment (benmccann) [inline pnpm-lock.yaml:7221]: think it'd be possible to upgrade `esbuild` to 0.19? that would reduce the number of versions of esbuild that we pull in

File: packages/adapter-cloudflare/package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
 	},
 	"dependencies": {
 		"@cloudflare/workers-types": "^4.20231121.0",
+		"wrangler": "^3.24.0",
```
Comment (benmccann) [inline packages/adapter-cloudflare/package.json:36]: Will users expect to install `wrangler` and control what version of `wrangler` is used? I'm wondering if it's better here or in `peerDependencies`. You probably know better than I do though

File: pnpm-lock.yaml (hunk 1 above)
Comment (dario-piotrowicz) [reply pnpm-lock.yaml:7221]: No unfortunately we can't upgrade `esbuild` to 0.19 😓 
See: https://github.com/evanw/esbuild/issues/3337

File: packages/adapter-cloudflare/package.json (hunk 2 above)
Comment (dario-piotrowicz) [reply packages/adapter-cloudflare/package.json:36]: Yes, when developing their app they will use `wrangler` anyways so I think that it makes sense to have it as a peer dependency, I've moved it 👍

But with this change, since `wrangler` might actually not be installed, I guess I might need to change the wrangler import for `getBindingsProxy` to be dynamic and try-catched so that we can check if it actually succeeds

What do you think?

Comment (benmccann) [reply packages/adapter-cloudflare/package.json]: Yes, putting it in a try-catch so we can print a nice error message when it is not present makes sense to me

Comment (benmccann) [reply packages/adapter-cloudflare/package.json]: Can you regenerate the pnpm lock file as well?

Comment (dario-piotrowicz) [reply packages/adapter-cloudflare/package.json]: I did re-run `pnpm i` but it didn't generate any different in the pnpm lock file, I guess it should? 😕

File: packages/kit/src/exports/public.d.ts [hunk 3]
```diff
@@ -45,6 +45,11 @@ export interface Adapter {
 		 */
 		read?: (details: { config: any; route: { id: string } }) => boolean;
 	};
+	/**
+	 * This function is used to allow the adapter to emulate the platform object during dev and
+	 * preview
+	 */
+	emulatePlatform?(svelteKitConfig: KitConfig): MaybePromise<App.Platform>;
```
Comment (Rich-Harris) [inline packages/kit/src/exports/public.d.ts:52]: I don't _think_ we need to pass the config into here, unless you had something specific in mind? If we do need it, it should be `opts: { config: KitConfig }` so that we can add other stuff in future if necessary

```suggestion
	emulatePlatform?(): MaybePromise<App.Platform>;
```

**Comment by Rich-Harris** [comment]
Thinking about what the implementation will look like for some of that stuff, it's probably quite finicky and involves a lot of knowledge of the internals, so if you're cool with it @dario-piotrowicz then maybe it's more logical for us to pick it up from here? Appreciate all the work that you've put into it so far!

**Comment by Rich-Harris** [comment]
I've made a start in https://github.com/sveltejs/kit/pull/11730, though there are a few TODOs and open questions still outstanding

File: packages/kit/src/exports/public.d.ts (hunk 3 above)
Comment (dario-piotrowicz) [reply packages/kit/src/exports/public.d.ts:52]: It was just for future proofing the solution/allowing other adapters to add things on top of it (which your solution addresses better)

Anyways it is not needed and it can be omitted entirely if you prefer (and/or omitted for now and potentially added when/if needed)

**Comment by dario-piotrowicz** [comment]
@Rich-Harris Thanks so much for looking into this 🙂 

We on the Cloudflare team are really keen and hopeful to get this sort of dev bindings support ready as soon as we can. We're also working with other framework authors to apply the same across the whole js ecosystem.
Whatever you think it's best, both for your framework and for hopefully expedite this feature's implementation works for us, and we're always happy to keep working on this and/or support in any other way we can.

So yeah please feel free to take it over if you want, but I am always here to pick it back up and/or help in any manner I can.

**Comment by Rich-Harris** [comment]
Thank you @dario-piotrowicz! If you have any pointers on how to populate `caches` and `context`, and how to test it locally (see notes on the other PR), we'd be very grateful. Also, is it okay that we're not calling `bindings.dispose` but are just killing the process?

**Comment by saturnonearth** [comment]
Probably a useless comment here that you are all already aware of, but [wrangler-proxy](https://github.com/chientrm/wrangler-proxyl) is a great example of how to use wrangler in development for SK - but I thought it might help 🤷‍♂️
